sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -858,6 +859,30 @@ object PushPredicateThroughJoin extends Rule[LogicalPlan] with PredicateHelper {
 }
 
 /**
+  * Push [[Limit]] operators through [[Join]] operators, iff the join type is outer joins.
+  * Adding extra [[Limit]] operators on top of the outer-side child/children.
+  */
+object PushLimitThroughOuterJoin extends Rule[LogicalPlan] with PredicateHelper {
+
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case f @ Limit(expr, Join(left, right, joinType, joinCondition)) =>
+      joinType match {
+        case RightOuter =>
+          Limit(expr, Join(left, CombineLimits(Limit(expr, right)), joinType, joinCondition))
```

Comment:
Since we call `CombineLimits` here to combine two consecutive `Limit`s, we will not add extra `Limit` in the subsequent iteration. Thus, no change will be made in the plan. Thus, `RuleExecutor` will stop automatically. Am I right?